pkg/controller/ports/assignment/managed.go

```diff
@@ -0,0 +1,169 @@
+)
+
+// SetupAssignment adds a controller that reconciles Assignments
+func SetupAssignment(mgr ctrl.Manager, l logging.Logger) error {
+	name := managed.ControllerName(v1alpha1.AssignmentGroupKind)
+
+	r := managed.NewReconciler(mgr,
+		resource.ManagedKind(v1alpha1.AssignmentGroupVersionKind),
+		managed.WithExternalConnecter(&connecter{kube: mgr.GetClient()}),
+		managed.WithInitializers(),
+		managed.WithConnectionPublishers(),
+		managed.WithReferenceResolver(managed.NewAPISimpleReferenceResolver(mgr.GetClient())),
+		managed.WithLogger(l.WithValues("controller", name)),
+		managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name))),
+	)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1alpha1.Assignment{}).
+		Complete(r)
+}
+
+type connecter struct {
+	kube        client.Client
+	newClientFn func(ctx context.Context, credentials []byte, projectID string) (portsclient.ClientWithDefaults, error)
+}
+
+func (c *connecter) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	g, ok := mg.(*v1alpha1.Assignment)
+	if !ok {
+		return nil, errors.New(errNotAssignment)
+	}
+
+	p := &packetv1alpha2.Provider{}
+	n := meta.NamespacedNameOf(g.Spec.ProviderReference)
+	if err := c.kube.Get(ctx, n, p); err != nil {
+		return nil, errors.Wrap(err, errGetProvider)
+	}
+
+	if p.GetCredentialsSecretReference() == nil {
+		return nil, errors.New(errProviderSecretNil)
+	}
+
+	s := &corev1.Secret{}
+	n = types.NamespacedName{Namespace: p.Spec.CredentialsSecretRef.Namespace, Name: p.Spec.CredentialsSecretRef.Name}
+	if err := c.kube.Get(ctx, n, s); err != nil {
+		return nil, errors.Wrap(err, errGetProviderSecret)
+	}
+	newClientFn := portsclient.NewClient
+	if c.newClientFn != nil {
+		newClientFn = c.newClientFn
+	}
+	client, err := newClientFn(ctx, s.Data[p.Spec.CredentialsSecretRef.Key], p.Spec.ProjectID)
+
+	return &external{kube: c.kube, client: client}, errors.Wrap(err, errNewClient)
+}
+
+type external struct {
+	kube   client.Client
+	client portsclient.ClientWithDefaults
+}
+
+func (e *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	a, ok := mg.(*v1alpha1.Assignment)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotAssignment)
+	}
+
+	// Observe port
+	port, err := e.client.GetPortByName(a.Spec.ForProvider.DeviceID, a.Spec.ForProvider.Name)
+	if packetclient.IsNotFound(err) {
+		return managed.ExternalObservation{}, errors.New("port does not exist")
+	}
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(err, errGetPort)
+	}
+
+	o := managed.ExternalObservation{
+		ResourceExists:   false,
+		ResourceUpToDate: true,
+	}
+
+	for _, net := range port.AttachedVirtualNetworks {
+		if net.ID == a.Spec.ForProvider.VirtualNetworkID {
+			a.Status.SetConditions(runtimev1alpha1.Available())
+			resource.SetBindable(a)
+			o.ResourceExists = true
+		}
+	}
+
+	meta.SetExternalName(a, port.ID)
```

Comment:
This is admittedly a bit of a weird way to use the `ExternalName` annotation here. It is not used in the `Observe` call at all and the assignment itself is not actually an API object on the packet side. However, once we set it we don't have to make `GetPortByName` calls in `Create` and `Delete` anymore. We could just as easily put this into the `Status` and reference it in that way.

Reply:
It makes sense. Maybe Href would be a better field to use, but it doesn't make much difference (and Href would need more handling to use).